### PR TITLE
add test for trying to create BitArray from infinite iter

### PR DESCRIPTION
**Pull request by git-john-gardner**
_[My first PR]_ 
Improve the [code coverage](https://coveralls.io/builds/42489547/source?filename=base%2Fbitarray.jl#L602) by testing that the `BitArray` constructor throws a `MethodError` when passing an iterator with `IsInfinite` size.

### CODE REVIEW

File: test/bitarray.jl [hunk 1]
```diff
@@ -203,6 +203,15 @@ timesofar("utils")
         @test_throws DimensionMismatch BitMatrix((isodd(i) for i in 1:3))
     end
 
+    @testset "constructor from infinite iterable" begin
+        struct InfiniteIter; state::Bool; end
```
Comment (git-john-gardner) [inline test/bitarray.jl:207]: I was unsure whether using an existing infinite sized iter here would have been better here, e.g. `Base.Iterators.countfrom` - this is simpler, but adds dependence on `Base.Iterators`...

File: test/bitarray.jl [hunk 2]
```diff
@@ -203,6 +203,15 @@ timesofar("utils")
...
+        Base.iterate(i::InfiniteIter, state=1) = (state, !state)
+        Base.IteratorSize(i::InfiniteIter) = Base.IsInfinite()
```
Comment (sostock) [inline test/bitarray.jl:209]:
```suggestion
        Base.IteratorSize(::Type{InfiniteIter}) = Base.IsInfinite()
```
Traits like `Base.IteratorSize` should be defined for types, not their instances.

File: test/bitarray.jl (hunk 1 above)
Comment (sostock) [inline test/bitarray.jl:207]:
```suggestion
        struct InfiniteIter end
```
The struct doesn’t need a field, it is never used.

File: test/bitarray.jl [hunk 3]
```diff
@@ -203,6 +203,15 @@ timesofar("utils")
         @test_throws DimensionMismatch BitMatrix((isodd(i) for i in 1:3))
     end
 
+    @testset "constructor from infinite iterable" begin
+        struct InfiniteIter; state::Bool; end
+        Base.iterate(i::InfiniteIter, state=1) = (state, !state)
```
Comment (sostock) [inline test/bitarray.jl:208]:
```suggestion
        Base.iterate(i::InfiniteIter, state=true) = (state, !state)
```
Using `1` (an `Int`) as argument will not work:
```
julia> !1
ERROR: MethodError: no method matching !(::Int64)
```

**Review by sostock — COMMENTED** [review]
Thanks for your contribution!

I think there is nothing wrong with using `Base.Iterators` here (I would actually prefer it over defining a new type just for a test). If you want to keep the `InfiniteIter` implementation, there are some issues with it that would need to be changed.

**Comment by git-john-gardner** [comment]
Thanks for the suggestions @sostock! I have changed the PR to use a predefined iterator.

**Comment by KristofferC** [comment]
Thanks and sorry for the slow merge.

**Pull request merged**